pkg/cmd/roachtest/operation_impl.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/operation"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+var errOperationFatal = errors.New("o.Fatal() was called")
+
+type operationImpl struct {
+	spec      *registry.OperationSpec
+	cockroach string // path to main cockroach binary on the cluster.
+
+	// l is the logger that the operation will use for its output.
+	l *logger.Logger
+
+	mu struct {
+		syncutil.RWMutex
+		done bool
+
+		// cancel, if set, is called from the o.Fatal() family of functions when the
+		// op is being marked as failed (i.e. when the failures slice is being
+		// appended to). This is used to cancel the context passed to o.spec.Run(),
+		// so async goroutines can be notified.
+		cancel func()
+
+		// failures added via addFailures, in order. An operation may have multiple
+		// calls to o.Fail()/Error(), with each call adding to this slice once.
+		failures []error
+
+		status string
+	}
+	cleanupState map[string]string
+}
+
+func (o *operationImpl) ClusterCockroach() string {
+	return o.cockroach
+}
+
+func (o *operationImpl) Name() string {
+	return o.spec.Name
+}
+
+// L returns the operation's logger.
+func (o *operationImpl) L() *logger.Logger {
+	return o.l
+}
+
+func (o *operationImpl) status(ctx context.Context, args ...interface{}) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.mu.status = fmt.Sprint(args...)
```

Comment:
Why do we need a `status` field?

Reply:
I use this in drt-run to grab the "current" status of an operation. Could drop it if we don't do the clusterImpl work we talked about earlier and go with the subprocess approach, but worth keeping for now.